Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide/Internal.hs

```diff
@@ -90,15 +96,36 @@ endRideTransaction ::
   Id DP.Driver ->
   SRB.Booking ->
   Ride.Ride ->
+  Bool ->
+  LatLong ->
   Maybe DFare.FareParameters ->
   Maybe (Id RD.RiderDetails) ->
   DFare.FareParameters ->
   TransporterConfig ->
   Id Merchant ->
-  m ()
-endRideTransaction driverId booking ride mbFareParams mbRiderDetailsId newFareParams thresholdConfig merchantId = do
+  m (Maybe DLocation.Location)
+endRideTransaction driverId booking ride pickupDropOutsideOfThreshold tripEndPoint mbFareParams mbRiderDetailsId newFareParams thresholdConfig merchantId = do
   triggerRideEndEvent RideEventData {ride = ride{status = Ride.COMPLETED}, personId = cast driverId, merchantId = merchantId}
   triggerBookingCompletedEvent BookingEventData {booking = booking{status = SRB.COMPLETED}, personId = cast driverId, merchantId = merchantId}
+  endLocationCustomer <-
+    if pickupDropOutsideOfThreshold
+      then do
+        toLocationResponse <- Maps.getTripPlaceName merchantId GetPlaceNameReq {getBy = ByLatLong LatLong {lat = tripEndPoint.lat, lon = tripEndPoint.lon}, sessionToken = Nothing, language = Nothing}
+        actualToLocationId <- generateGUID
+        toAddress <- GoogleMaps.mkLocation toLocationResponse
+        actualToLocation <- buildLocation actualToLocationId tripEndPoint toAddress
+        mappers <- QLocationMapping.findByTagId $ ride.id.getId
+        let toLocationMappers = filter (\mapper -> mapper.order /= 0) mappers
+        logDebug $ "updating location mappings " <> show toLocationMappers
+        logDebug $ "trip end point is " <> show actualToLocation
+        epochVersion <- liftIO $ round `fmap` Time.getPOSIXTime
+        let epochVersionLast5Digits = epochVersion `mod` 100000 :: Integer
+        Esq.runTransaction $ QLocation.create actualToLocation
+        for_ toLocationMappers $ \locMap -> do
+          Esq.runTransaction $ QLocationMapping.updateLocationInMapping locMap actualToLocation epochVersionLast5Digits
```

Comment:
Why do we need these separate transactions inside of cycle? Can we put all write queries to transaction on line 150?